mullvad-relay-selector/src/relay_selector/matcher.rs

```diff
@@ -0,0 +1,282 @@
+    pub locations: Constraint<ResolvedLocationConstraint>,
+    /// Relay providers allowed to be picked from.
+    pub providers: Constraint<Providers>,
+    /// Relay ownership allowed to be picked from.
+    pub ownership: Constraint<Ownership>,
+    /// Concrete representation of [`RelayConstraints`] or [`BridgeConstraints`].
+    pub endpoint_matcher: T,
+}
+
+impl RelayMatcher<AnyTunnelMatcher> {
+    pub fn new(
+        constraints: RelayQuery,
+        openvpn_data: OpenVpnEndpointData,
+        brige_state: BridgeState,
+        wireguard_data: WireguardEndpointData,
+        custom_lists: &CustomListsSettings,
+    ) -> Self {
+        Self {
+            locations: ResolvedLocationConstraint::from_constraint(
+                constraints.location,
+                custom_lists,
+            ),
+            providers: constraints.providers,
+            ownership: constraints.ownership,
+            endpoint_matcher: AnyTunnelMatcher {
+                wireguard: WireguardMatcher::new(constraints.wireguard_constraints, wireguard_data),
+                openvpn: OpenVpnMatcher::new(
+                    constraints.openvpn_constraints,
+                    openvpn_data,
+                    brige_state,
+                ),
+                tunnel_type: constraints.tunnel_protocol,
+            },
+        }
+    }
+}
+
+impl<T: EndpointMatcher> RelayMatcher<T> {
+    /// Filter a list of relays and their endpoints based on constraints.
+    /// Only relays with (and including) matching endpoints are returned.
+    pub fn filter_matching_relay_list<'a, R: Iterator<Item = &'a Relay> + Clone>(
+        &self,
+        relays: R,
+    ) -> Vec<Relay> {
+        let shortlist = relays
+            // Filter on active relays
+            .filter(|relay| filter_on_active(relay))
+            // Filter by location
+            .filter(|relay| filter_on_location(&self.locations, relay))
+            // Filter by ownership
+            .filter(|relay| filter_on_ownership(&self.ownership, relay))
+            // Filter by providers
+            .filter(|relay| filter_on_providers(&self.providers, relay))
+            // Filter on relay type & relay specific properties
+            .filter(|relay| self.endpoint_matcher.is_matching_relay(relay));
+
+        // The last filtering to be done is on the `include_in_country` attribute found on each
+        // relay. A regular, user-facing relay will have `include_in_country` set to true.
+        // If a relay has `include_in_country` set to false, they are purposely hidden than
+        // other relays. We should only consider those if there are no regular candidates left.
+        let ignore_include_in_country = !shortlist.clone().any(|relay| relay.include_in_country);
+        shortlist
+            .filter(|relay| {
+                self.locations
+                    .matches_with_opts(relay, ignore_include_in_country)
+            })
+            .cloned()
+            .collect()
```

Comment:
I find the matches_with_opts(relay, ignore_include_in_country) function to be a bit confusing. Since we have already filtered based on location above, the ignore_include_in_country argument only has the effect of negating the entire filter, which is not evident without digging into its implementation.

I suggest refactoring to use an if-statement, which is more readable with identical behaviour (correct me if I'm wrong)
```suggestion
        if !shortlist.clone().any(|relay| relay.include_in_country) {
            shortlist
                .filter(|relay| self.locations.matches_with_opts(relay, true))
                .cloned()
                .collect()
        } else {
            shortlist.cloned().collect()
        }
```
Then I suggest wrapping `self.locations.matches_with_opts(relay, true)` in a function called, say, `filter_on_include_in_country`, similar to how `filter_on_location` works.